src/zkp/whitelist.rs

```diff
@@ -0,0 +1,470 @@
+//! Bindings for the "whitelist" ring signature implementation in secp256k1-zkp.
+//!
+//! This implementation is used for Liquid PAK list inclusion proofs.
+
+#[cfg(feature = "std")]
+use std::{fmt, str};
+
+use core::ptr;
+
+use ffi::CPtr;
+#[cfg(feature = "std")]
+use from_hex;
+use {ffi, secp256k1, Error, PublicKey, Secp256k1, SecretKey, Signing, Verification};
+
+/// A whitelist ring signature.
+#[derive(Clone, PartialEq, Eq, Hash)]
+#[repr(transparent)]
+pub struct WhitelistSignature(ffi::WhitelistSignature);
+
+impl WhitelistSignature {
+    /// Number of keys in the whitelist.
+    pub fn n_keys(&self) -> usize {
+        self.0.n_keys as usize
+    }
+
+    /// Serialize to bytes.
+    #[cfg(feature = "std")]
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = vec![0; 33 + 32 * self.n_keys()];
+
+        let mut out_len = buf.len();
+        let ret = unsafe {
+            ffi::secp256k1_whitelist_signature_serialize(
+                ffi::secp256k1_context_no_precomp,
+                buf.as_mut_ptr(),
+                &mut out_len,
+                &self.0,
+            )
+        };
+        assert_eq!(ret, 1, "failed to serialize whitelist signature");
+        assert_eq!(
+            out_len,
+            buf.len(),
+            "whitelist serialized to unexpected length"
+        );
+
+        buf
+    }
+
+    /// Parse a whitelist ring signature from a byte slice.
+    pub fn from_slice(bytes: &[u8]) -> Result<Self, Error> {
+        let mut sig = ffi::WhitelistSignature::default();
+
+        let ret = unsafe {
+            ffi::secp256k1_whitelist_signature_parse(
+                ffi::secp256k1_context_no_precomp,
+                &mut sig,
+                bytes.as_ptr(),
+                bytes.len(),
+            )
+        };
+        if ret != 1 {
+            return Err(Error::InvalidWhitelistSignature);
+        }
+
+        Ok(WhitelistSignature(sig))
+    }
+
+    /// Create a new whitelist ring signature for the given PAK list and whitelist key.
+    pub fn sign<C: Signing>(
```

Comment:
Shall we name this `new` instead?

`WhitelistSignature::sign` repeats the `sign` part.

Reply:
For reference, `SurjectionProof` for example uses `new`: https://github.com/ElementsProject/rust-secp256k1-zkp/blob/0dba4c033a929164dd21c488d2a5a9790919f65c/src/zkp/surjection_proof.rs#L21-L31

Reply:
I would prefer `sign` if we are calling it on whatever we are signing. Like `message.sign(key)`. In this case, that is not really possible from what I can see on the parameter list.

I am just co-maintainer here so happy to go with whatever as long as we are consistent :)

Reply:
Changed.